### PR TITLE
prevent reject resolution reminders from being sent if another intake exists with the same hashed ssn and a submission

### PR DESCRIPTION
**Pull request by mrotondo**
## Link to pivotal/JIRA issue
- https://codeforamerica.atlassian.net/browse/FYST-2058
## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

**Reminder**: merge main into this branch and get green tests before merging to main
## What was done?
- This PR adds a check to `SendRejectResolutionReminderNotificationJob` to prevent the message from being sent if another intake exists with the same hashed SSN at at least one efile submission.
- There are other messages which will need the same treatment (described in https://codeforamerica.atlassian.net/browse/FYST-2057) but I'm pushing this up first to get it into acceptance so that we can figure out our acceptance process as well as find any corner cases where we want to potentially send the message despite the existence of such an intake.
## How to test?
- Automated testing of the job's new functionality has been added
- Manual testing would look like (per @mpidcock):
  - if testing in heroku, these would be the steps:
    - partially complete an intake for a persona (don't submit)
    - start a new intake with a different email or phone
    - can be the same login type or not
    - complete the same persona all the way through submission
    - run the reminder rake task
    - verify no message is sent

### CODE REVIEW

File: spec/jobs/state_file/send_reject_resolution_reminder_notification_job_spec.rb [hunk 1]
```diff
@@ -110,7 +111,36 @@
...
+
+          context "and which was filed in the same state" do
+            let(:other_intake_class) { :state_file_az_intake }
+
+            it "does not send the message" do
+              expect { described_class.perform_now(intake) }.not_to change(StateFileNotificationEmail, :count)
+            end
           end
+
+          context "and which was filed in another state" do
```
Comment (anisharamnani) [inline spec/jobs/state_file/send_reject_resolution_reminder_notification_job_spec.rb:135]: @mpidcock was saying that this wasn’t possible?

Comment (mrotondo) [reply spec/jobs/state_file/send_reject_resolution_reminder_notification_job_spec.rb:135]: it may indeed not be!

Comment (anisharamnani) [reply spec/jobs/state_file/send_reject_resolution_reminder_notification_job_spec.rb:135]: on second thought it is still good to test

File: spec/jobs/state_file/send_reject_resolution_reminder_notification_job_spec.rb [hunk 2]
```diff
@@ -110,7 +111,36 @@
               message: message,
               body_args: body_args).exactly(2).times
           end
+        end
+
+        context "when another intake exists with the same hashed SSN" do
```
Comment (anisharamnani) [inline spec/jobs/state_file/send_reject_resolution_reminder_notification_job_spec.rb:116]: can we add a test case where they have the same ssn but don’t have an efile submission?

File: app/jobs/state_file/send_reject_resolution_reminder_notification_job.rb [hunk 3]
```diff
@@ -24,6 +25,17 @@ def return_status_link(intake)
       self.class.return_status_link(intake.locale || "en")
     end
 
+    def other_intake_with_same_ssn_has_submission(intake)
+      return if intake.hashed_ssn.nil?
+      StateInformationService.state_intake_classes.any? do |intake_class|
```
[the remark this answers is not in the document]
Comment (mrotondo) [reply app/jobs/state_file/send_reject_resolution_reminder_notification_job.rb:30]: I'll add the NY exclusion, and it's working because we're already in the StateFile module here

[the remark this answers is not in the document]
Comment (mrotondo) [reply app/jobs/state_file/send_reject_resolution_reminder_notification_job.rb:30]: (though we should be free of worries about NY intakes, since they were deleted as part of the Get Your PDF work)

File: app/jobs/state_file/send_reject_resolution_reminder_notification_job.rb [hunk 4]
```diff
@@ -24,6 +25,17 @@ def return_status_link(intake)
       self.class.return_status_link(intake.locale || "en")
     end
 
+    def other_intake_with_same_ssn_has_submission(intake)
+      return if intake.hashed_ssn.nil?
```
[the remark this answers is not in the document]
Comment (mrotondo) [reply app/jobs/state_file/send_reject_resolution_reminder_notification_job.rb:29]: yeah, i agree that it's confusing  - return without an argument in ruby just returns `nil`. it doesn't have any non-local effects - this will just cause the method call on line 5 to evaluate to `nil`, which _will_ cause the job's `perform` to exit via an early return.

whether or not no-argument return is good or bad in general, in this case i agree with (and rubocop, now that i added the `?` to the method name) that this should be an explicit `return false`

[the remark this answers is not in the document]
Comment (mrotondo) [reply app/jobs/state_file/send_reject_resolution_reminder_notification_job.rb:29]: also, this job only runs for a specific intake - i know that's not true for all our messaging jobs, but in any case, `return` can't affect a job's invocation unless it's `return`ing from `perform` itself

File: app/jobs/state_file/send_reject_resolution_reminder_notification_job.rb [hunk 5]
```diff
@@ -24,6 +25,17 @@ def return_status_link(intake)
...
+      StateInformationService.state_intake_classes.any? do |intake_class|
+        intakes = intake_class
+          .where(hashed_ssn: intake.hashed_ssn)
+          .where.associated(:efile_submissions)
```
[the remark this answers is not in the document]
Comment (mrotondo) [reply app/jobs/state_file/send_reject_resolution_reminder_notification_job.rb:33]: neither did i! i was digging into different ways to do this, and found out this method was added in rails 7 (there's also `.where.missing` to do the opposite)

**Review by arinchoi03 — APPROVED** [review]
🛸

**Comment by arinchoi03** [comment]
@mrotondo you can merge in main & get the flipper + the refactor to move the method to the intake
